modules/marshaller-common/src/main/java/org/apache/ignite/internal/marshaller/Marshaller.java

```diff
@@ -64,16 +65,35 @@ public static <T> Marshaller createMarshaller(
      * @param mapper Mapper.
      * @return Marshaller.
      */
-    private static <T> SimpleMarshaller simpleMarshaller(MarshallerColumn[] cols, OneColumnMapper<T> mapper) {
-        final BinaryMode mode = MarshallerUtil.mode(mapper.targetType());
+    private static SimpleMarshaller simpleMarshaller(MarshallerColumn[] cols, OneColumnMapper<?> mapper) {
+        int colIdx = -1;
 
-        final MarshallerColumn col = cols[0];
+        if (mapper.mappedColumn() == null) {
+            if (cols.length != 1) {
+                throw new IllegalArgumentException(
+                        "Failed to map object to a single column, because schema contains more columns and no mapped columns were provided"
+                );
+            }
 
-        assert cols.length == 1;
-        assert mode == col.type() : "Target type is not compatible.";
-        assert !mapper.targetType().isPrimitive() : "Non-nullable types are not allowed.";
+            colIdx = 0;
+        } else {
+            for (int i = 0; i < cols.length; i++) {
+                if (cols[i].name().equals(mapper.mappedColumn())) {
+                    colIdx = i;
+                }
+            }
+
+            if (colIdx < 0) {
+                throw new IllegalArgumentException("Failed to map object to a single column:" + mapper.mappedColumn());
+            }
+        }
+
+        assert !mapper.targetType().isPrimitive() : "Non-nullable types are not allowed";
```

Comment:
Should we change it to `IllegalArgumentException`? This can happen because of the user code, right?